internal/auth/tokens.go

```diff
@@ -25,65 +27,147 @@ const (
 	GithubJWTAudience = "nscloud.dev/inline-token"
 )
 
+var ErrTokenNotExist = errors.New("token does not exist")
+
 type Token struct {
-	TenantToken string `json:"tenant_token,omitempty"`
+	Username    string   `json:"username,omitempty"`
+	Scopes      []string `json:"scopes,omitempty"`
+	TenantToken string   `json:"tenant_token,omitempty"`
 }
 
-func StoreTenantToken(token string) error {
-	data, err := json.Marshal(Token{TenantToken: token})
-	if err != nil {
-		return err
+func (t Token) Equal(other Token) bool {
+	if t.Username == other.Username {
+		slices.Sort(t.Scopes)
```

Comment:
A method like `Equals` should not mutate its state.

Either pre-sort, or clone the slice to sort.